### PR TITLE
Exec: Add ability to set environment variables

### LINKED ISSUE #24355: Ability to set environment variables in exec
Currently it's only supported on run

(working on a patch)

### PR DESCRIPTION
**Pull request by jwendell**
Keeping the current behavior for exec, i.e., inheriting
variables from main process. New variables will be added
to current ones. If there's already a variable with that
name it will be overwritten.

Example of usage: docker exec -it -e TERM=vt100 <container> top

Closes #24355.

Signed-off-by: Jonh Wendell jonh.wendell@redhat.com

### CODE REVIEW

**Comment by cyphar** [comment]
Can you link the relevant `engine-api` and `containerd` PRs?

**Comment by jwendell** [comment]
engine-types: https://github.com/docker/engine-api/pull/323

about containerd: I touched libcontainerd/ directory. Are you referring to that? If yes, what's its upstream repo?

**Comment by cyphar** [comment]
My apologies, you just modified what we send as `spec.Env`. There isn't a matching PR necessary.

**Comment by runcom** [comment]
This was discussed so many times and we always closed PR like this:

https://github.com/docker/docker/pull/16794
https://github.com/docker/docker/pull/14202#issuecomment-123893860

And the roadmap:

https://github.com/docker/docker/blob/master/ROADMAP.md#21-docker-exec

**Comment by thaJeztah** [comment]
We discussed this PR in a review session, and while we still don't want to overload `docker exec` with too many options, most maintainers present thought having an `-e` / `--env` option would be useful for situations where you need to set an environment variable for the process `exec`'d, but don't have a shell in the container.

Given that this was rejected a few times in the past, I don't want to move this forward to "code review" directly, so ping @docker/core-engine-maintainers PTAL

If there are no strong objections, I think we can move this to code review in a week or so (feel free to ping me if I forget) :smile:

/cc @tianon @tiborvass @tonistiigi @jhowardmsft @cpuguy83 @justincormack

**Comment by duglin** [comment]
SGTM 
but we'll need testcases with the code.

**Comment by crosbymichael** [comment]
sgtm for adding --env,e to exec

**Comment by jhowardmsft** [comment]
I don't have a strong objection, so SGTM too.

**Comment by vdemeester** [comment]
sgtm too, it definitely can help some time 😇

**Comment by mlaventure** [comment]
SGTM, just hit a case where I needed this :)

An integration test is needed though.

**Comment by vdemeester** [comment]
Let's move this to code review then. @jwendell it needs a rebase, a PR in `engine-api` and some integration test(s) :angel:

**Comment by jwendell** [comment]
ok, I'll do this later today ;)

**Comment by jwendell** [comment]
engine pr: https://github.com/docker/engine-api/pull/323

**Comment by cpuguy83** [comment]
Needs a rebase/re-vendor for engine-api changes.

**Comment by cpuguy83** [comment]
Also needs an integration test. Test should make sure the env from the container is propagated to the exec as well as getting the new envs.

**Comment by cpuguy83** [comment]
And now engine-api is moved to `github.com/docker/docker/api/types`, no need to vendor.

**Comment by vdemeester** [comment]
ping @jwendell

**Comment by jwendell** [comment]
OK, will fix this today.

**Comment by justincormack** [comment]
@jwendell any update?

**Comment by runcom** [comment]
@jwendell ping, could you rebase this patch on latest master?

**Comment by jwendell** [comment]
doing this now

**Comment by jwendell** [comment]
Rebased and pushed, but I have no idea why the tests failed, it seems unrelated do the changes.

**Comment by thaJeztah** [comment]
Janky and Experimental fail on this; could be related, although not sure what's the reason;

```
20:50:54 ----------------------------------------------------------------------
20:50:54 FAIL: docker_cli_exec_unix_test.go:85: DockerSuite.TestExecWithNoTERM
20:50:54 
20:50:54 docker_cli_exec_unix_test.go:91:
20:50:54     c.Assert(err, checker.IsNil)
20:50:54 ... value *exec.ExitError = &exec.ExitError{ProcessState:(*os.ProcessState)(0xc42108dae0), Stderr:[]uint8(nil)} ("exit status 1")
20:50:54 
20:50:55 
20:50:55 ---
```

**Comment by thaJeztah** [comment]
WindowsRS1 failure is strange; have to look into that (looks legitimate, but not related to this PR)

```
20:41:48 ----------------------------------------------------------------------
20:41:48 FAIL: docker_cli_build_test.go:6549: DockerSuite.TestBuildLabelsOverride
20:41:48 
20:41:48 docker_cli_build_test.go:6641:
20:41:48     c.Assert(err, check.IsNil)
20:41:48 ... value *errors.errorString = &errors.errorString{s:"failed to build the image: Sending build context to Docker daemon 2.048 kB\r\r\nStep 1/2 : FROM scratch\nWindows does not support FROM scratch\n"} ("failed to build the image: Sending build context to Docker daemon 2.048 kB\r\r\nStep 1/2 : FROM scratch\nWindows does not support FROM scratch\n")
20:41:48 
```

**Comment by thaJeztah** [comment]
Windows will be fixed by https://github.com/docker/docker/pull/27416; looks like two PRs got merged, which both succeeded individually, but failed after both were merged

**Comment by jwendell** [comment]
green bar

**Comment by crosbymichael** [comment]
LGTM

**Comment by runcom** [comment]
LGTM moving to Doc review

**Comment by runcom** [comment]
Ping @thaJeztah @vdemeester for doc review

File: docs/reference/api/docker_remote_api_v1.25.md
Comment (thaJeztah) [inline docs/reference/api/docker_remote_api_v1.25.md]: Can you change the "form" example to:

```Markdown
`["VAR=value", ...]`
```

The other format was confusing some people (because they interpreted the square brackets as being part of the JSON)

File: docs/reference/api/docker_remote_api_v1.25.md
Comment (thaJeztah) [inline docs/reference/api/docker_remote_api_v1.25.md]: (nitpick), can you change the indenting to match other examples (we're not completely consistent yet, but trying to make it more consistent with each PR :smile:);

```patch
diff --git a/docs/reference/api/docker_remote_api_v1.25.md b/docs/reference/api/docker_remote_api_v1.25.md
index 5b29ef5..efd9114 100644
--- a/docs/reference/api/docker_remote_api_v1.25.md
+++ b/docs/reference/api/docker_remote_api_v1.25.md
@@ -3149,9 +3149,9 @@ Sets up an exec instance in a running container `id`
       "Cmd": ["sh"],
       "DetachKeys": "ctrl-p,ctrl-q",
       "Env": [
-                   "FOO=bar",
-                   "BAZ=quux"
-             ],
+        "FOO=bar",
+        "BAZ=quux"
+     ],
       "Privileged": true,
       "Tty": true,
       "User": "123:456"
```

File: docs/reference/commandline/exec.md
Comment (thaJeztah) [inline docs/reference/commandline/exec.md]: Can you update the man-page as well? https://github.com/docker/docker/blob/master/man/docker-exec.1.md

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks @jwendell, I left some comments inline. Also;
- can you add a note to the API changes; https://github.com/docker/docker/blob/master/docs/reference/api/docker_remote_api.md#v125-api-changes
- it would be great to have the bash and zsh completion scripts updated for this (https://github.com/docker/docker/tree/master/contrib/completion/bash, https://github.com/docker/docker/tree/master/contrib/completion/zsh), but let me know if you want to tackle those yourself, otherwise I'll ask @albers and @sdurrheimer to help out (and do it in a follow up)

**Comment by jwendell** [comment]
done with the requested docs changes

**Review by thaJeztah — APPROVED** [review]
perfect! LGTM

**Comment by thaJeztah** [comment]
ping @vdemeester @mstanleyjones PTAL

**Review by vdemeester — APPROVED** [review]
LGTM 🐮

**Pull request merged**
